src/lib/reasoners/theory.ml

```diff
@@ -830,14 +722,14 @@ module Main_Default : S = struct
 
       let gamma, _ = CC_X.assume_literals t.gamma [] facts in
       (* update to optimize with the new gamma *)
-      let objectives, reset_cs_env =
-        update_objectives t.objectives assumed gamma in
+      (* let objectives, reset_cs_env =
+         update_objectives t.objectives assumed gamma in *)
```

Comment:
Why is this no longer needed? What happens if an assertion renders an existing objective invalid?

Reply:
The purpose of `update_objectives` function was to ensure we add new optimization discovered while exploring a term. At this time, we could have optimization everywhere in formulae. In particular, it means the theory module could ignore an optimization constraint before the underlying formula was involved in a formula of the current trail of the SAT solver. I didn't check, but I guess we could even forget an optimization constraint by backtracking.

If we discover a new constraint, we have to throw away all the case splits. A flag in `update_objectives` was used to remember that and we call `reset_case_split_env` if the flag is up.

Now, we feed the SAT solver with all the optimization constraints at once.  I thought it was fine but in fact we perform case splits after each assume. Thus we have to reset casesplits after each `optimize` in `Theory`.

Reply:
Actually I did! I just forget to remove the dead code commented. :) Thanks for the remark. I will add a comment in `optimize`.

Reply:
Uhm I thought it was fine because I use the same semantic as the assert but I probably miss something in `solving_loop`. Thanks for the test.

Reply:
Ah dang, I take it that means it does not work then :/ I think it is probably an issue in the `Satml_frontend` / `Satml` rather than `solving_loop`. The way `push` and `pop` work in the Satml solver is that we create guards, and assertions within a push/pop assertions are, well, guarded: if the guard is false, the assertion is just ignored.

So for instance the following:

```smt-lib
(push 1)
(assert f1)
(check-sat)
```

is treated as if we had:

```smt-lib
(declare-const guard Bool)
(assert (=> guard f1))
(assert guard) ; Generated for all assertions levels that were not popped on (check-sat)
(check-sat)
```

and the following:

```smt-lib
(push 1)
(assert f1)
(pop 1)
(check-sat)
```

is treated as if we had:

```smt-lib
(declare-const guard Bool)
(assert (=> guard f1))
(assert (not guard)) ; Generated on (pop 1)
(check-sat)
```

and so `Th.assume` (which is called during propagation) will never be called on `f1`. But I believe that with the current implementation (from my recollection of it, I did not check in detail) we will always call `optimize` without taking the guards into account.

In the previous implementation this "just worked" because optimization markers were expressions inside assertions that were not `assume`d when the guards are false, but now that optimization markers are declarations there is a little bit more work that needs to be duplicated. In fact there is probably a bit of overlap with #901, now that I think about it.

Given that optimization is still experimental (and that you've said you're sick of working on it for so long, which is understandable :) ), I think dropping support for optimization within push/pop in this PR is acceptable. We can then either deal with that as part of #901 (which I think is on track for inclusion in the next release, since postponing #904 freed up quite some time for me) or otherwise in a separate PR; your call.

Reply:
I think I take of it after we solve #901. I'll print a warning if we try to optimize with `push` and `pop`. An error isn't necessary here as we still got a correct model but we don't respect the optimization constraint.